mypyc/lib-rt/str_ops.c

```diff
@@ -43,18 +43,89 @@ PyObject *CPyStr_GetItem(PyObject *str, CPyTagged index) {
     }
 }
 
-PyObject *CPyStr_Build(int len, ...) {
-    int i;
+// A simplification of _PyUnicode_JoinArray() from CPython 3.9.6
+PyObject *CPyStr_Build(Py_ssize_t len, ...) {
+    Py_ssize_t i;
     va_list args;
+
+    // Calculate the total amount of space and check
+    // whether all components have the same kind.
+    Py_ssize_t sz = 0;
+    Py_UCS4 maxchar = 0;
+    int use_memcpy = 1; // Use memcpy by default
+    PyObject *last_obj = NULL;
+
     va_start(args, len);
+    for (i = 0; i < len; i++) {
+        PyObject *item = va_arg(args, PyObject *);
+        if (!PyUnicode_Check(item)) {
+            PyErr_Format(PyExc_TypeError,
+                         "sequence item %zd: expected str instance,"
+                         " %.80s found",
+                         i, Py_TYPE(item)->tp_name);
+            return NULL;
+        }
+        if (PyUnicode_READY(item) == -1)
+            return NULL;
 
-    PyObject *res = PyUnicode_FromObject(va_arg(args, PyObject *));
-    for (i = 1; i < len; i++) {
-        PyObject *str = va_arg(args, PyObject *);
-        PyUnicode_Append(&res, str);
-    }
+        Py_ssize_t add_sz = PyUnicode_GET_LENGTH(item);
+        Py_UCS4 item_maxchar = PyUnicode_MAX_CHAR_VALUE(item);
+        maxchar = Py_MAX(maxchar, item_maxchar);
 
+        if (add_sz + sz > (Py_ssize_t)PY_SSIZE_T_MAX) {
```

Comment:
This seems incorrect, because of possible overflow on the left hand side. Maybe this should be `add_sz > (size_t)(PY_SSIZE_T_MAX - sz)`? Also add comment about this.